src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs

```diff
@@ -0,0 +1,364 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+using Raven.Abstractions.Data;
+using Raven.Abstractions.Exceptions.Subscriptions;
+using Raven.Abstractions.Logging;
+using Raven.Server.Json;
+using Raven.Server.Routing;
+using Raven.Server.ServerWide.Context;
+using Voron.Data.Tables;
+using Sparrow.Json.Parsing;
+using Sparrow.Json;
+using System.Net.WebSockets;
+using System.Diagnostics;
+using System.Threading;
+using System.IO;
+using System.Net.Http.Headers;
+using Raven.Abstractions.Util;
+using Raven.Server.Extensions;
+using Sparrow;
+using Sparrow.Binary;
+
+
+namespace Raven.Server.Documents.Handlers
+{
+    public class SubscriptionsHandler : DatabaseRequestHandler
+    {
+
+        private static readonly ILog log = LogManager.GetLogger(typeof(SubscriptionsHandler));
+
+        [RavenAction("/databases/*/subscriptions/create", "POST", "/databases/{databaseName:string}/subscriptions/create")]
+        public async Task Create()
+        {
+            DocumentsOperationContext context;
+            using (ContextPool.AllocateOperationContext(out context))
+            {
+                var subscriptionCriteriaRaw = await context.ReadForDiskAsync(RequestBodyStream(), null).ConfigureAwait(false);
+                var subscriptionId = Database.SubscriptionStorage.CreateSubscription(subscriptionCriteriaRaw);
+                var ack = new DynamicJsonValue
+                {
+                    ["Id"] = subscriptionId
+                };
+                using (var writer = new BlittableJsonTextWriter(context, ResponseBodyStream()))
+                {
+                    context.Write(writer, ack);
+                }
+
+                HttpContext.Response.StatusCode = 201; // NoContent
+            }
+        }
+
+        [RavenAction("/databases/*/subscriptions", "DELETE",
+            "/databases/{databaseName:string}/subscriptions?id={subscriptionId:long}")]
+        public Task Delete()
+        {
+            var ids = HttpContext.Request.Query["id"];
+            if (ids.Count == 0)
+                throw new ArgumentException("The 'id' query string parameter is mandatory");
+
+            long id;
+            if (long.TryParse(ids[0], out id) == false)
+                throw new ArgumentException("The 'id' query string parameter must be a valid long");
+
+
+            Database.SubscriptionStorage.DeleteSubscription(id);
+
+            HttpContext.Response.StatusCode = 204; // NoContent
+
+            return Task.CompletedTask;
+        }
+
+        [RavenAction("/databases/*/subscriptions/open", "POST",
+            "/databases/{databaseName:string}/subscriptions/open?id={subscriptionId:long}")]
+        public Task Open()
```

Comment:
Why do we have separate `open` and `pull` requests? A pull / open is effectively the same, and it avoids having two separate requests